### ISSUE TITLE
Command PROGRAM.INFO

### ISSUE THREAD

**Issue by mougino** [RFO-BASIC/Basic#217]
Following discussion in [this forum thread](http://rfobasic.freeforums.org/current-program-being-run-t4548.html)

Add a new command:

> **Program.Info nexp | nvar**
> Returns information about the current program being run in a Bundle. If you provide a variable that is not a valid Bundle pointer, the command creates a new Bundle and returns the Bundle pointer in your variable. Otherwise it writes into the Bundle your variable or expression points to.

First set of bundle keys being returned;
- BasPath
- BasName
- SysPath
- UserApk

See full description ready to be copied to manual [here](http://rfobasic.freeforums.org/post32983.html#p32983)

**Comment by jMarcS** [comment]
This is going into v01.91 as is. I like what you have done with `Program.info`, including your effort to resist bloating it. Thank you for writing an entry for the Manual!

At some future time, we should consider another change.
If you run from the Editor without saving, `running_bas` (the same variable that feeds `Program.info BasName`) is empty. This makes `Run` exit with a syntax error. Ideally we would re-run the program in the Editor, but at the very least we should come up with a more meaningful error.

**Comment by olimaticer** [comment]
I have be made in my implementation something wrong, because unsaved
```
REM Start of BASIC! Program
Print "hallo world!"
Program.info b
bundle.get b , "BasName", pn$
Print "&&"+pn$+"**"
```
returns &&**, but no error. 

Some suggestion from my test version:

- Adding "PackageName"
- Maybe deleting "BasPath" and "SysPath", because implemented in file.root path$ , "_Basic"
- Writing "_PackageName" instead of "PackageName", so you see on the first look it is a system key or system value

Gregor

**Comment by jMarcS** [comment]
That's not the problem Gregor. The error comes from a `Run` command with no parameters:
````
Print "hallo world!"
Program.info b
bundle.get b , "BasName", pn$
Print "&&"+pn$+"**"
RUN
````
In v01.90.02, this ends the program with `SyntaxError`.
What does your implementation do?

**Comment by olimaticer** [comment]
Sorry Marc,

I forgot `RUN`.

For unsaved code with RUN or RUN"" I get the Message:

 "! Shortcut Error: \"" + aFileName + "\" not found"; --> from AutoRun.java
From my special device:
 ! Shortcut Error: "file:///storage/emulated/0/tfo-basic/source" not found;
This is an error, but with no interrupt.

I implement RUN without the first element, because Nicolas suggested

#218: Improve the RUN command to restart the current program if called without
(first) argument.
So in my implementation RUN equals RUN ""

In Run.java

```
	private boolean executeRUN() {
	// get program filename
		String fileName = "";//vv2017-02-23gt
		if (getStringArg())							// get program filename
			fileName = StringConstant;
		if (fileName.equals(""))
			fileName = "file://" + Editor.ProgramPath + "/" + Editor.ProgramFileName;
		String data = "";
		if (isNext(',')) {													// optional
			if (!getStringArg())		return false;						// parameter to pass to program
			data = StringConstant;
		}
		if (!checkEOL())				return false;
		String path = "";
		path = Basic.getFilePath(Basic.SOURCE_DIR, fileName);
		if (fileName.toLowerCase().indexOf("file://") == 0)path = fileName.substring(7);
		boolean exists = false;
		if (!Basic.isAPK) { exists = new File(path).exists(); }				// standard BASIC can only RUN a file
		else if (Basic.getRawResourceID(fileName) != 0) { exists = true; }	// APK can run resource
		else {																// or asset
			String assetPath = Basic.getAppFilePath(Basic.SOURCE_DIR, fileName);
			if (getAssetType(assetPath) == "f") { exists = true; }			// it is a valid asset file
		}
		if (!exists) {														// error if the program does not exist
			return RunTimeError(fileName + " not found");
		}
		// This AutoRun will be used later to load the new program
		// and to get an Intent to create a new Interpreter to run it.
		Context context = getApplicationContext();
		mAutoRun = new AutoRun(context, fileName, true, data);				// use file name without path
		Exit = true;														// do not allow interrupt processing
		return true;
	}
```

+++++++++++++++++++++++++++++++++++++++++++++++++


In AutoRun.java
```
					baseDriveChanged = true;
					//aFileName = " fubar";
				}
			}
			if (aFileName.toLowerCase().indexOf("file://") == 0)aFileName = aFileName.substring(7);//2017-02-23gt
		}

		Basic.clearProgram(); 				// Clear the old program
		Editor.DisplayText = "";			// Clear the display text buffer
		Run.PrintShow("");//2017-02-23 Without, BASIC! crashes.
		//Maybe deleting two wrong lines (mDataBundle) on top solves the issue, too.
		ArrayList<String> lines = new ArrayList<String>();
		int size = Basic.loadProgramFileToList(!mFromRun, aFileName, lines);
		Boolean found = (size != 0);		// size is 0 if file not found
		if (!found) {
			// Special case of file not found and doing Launcher Shortcut.
			// Turn the program file into an error message and act as if we loaded a file.
			String msg = baseDriveChanged
				? "! Shortcut created with different base drive."
				: "! Shortcut Error: \"" + aFileName + "\" not found";
			lines.add(msg);
			size = msg.length() + 1;
		}

```
I have to say, that I use a **public** Run.PrintShow() for **old fashioned** debugging.
( I belive the issue could be in Editor.java onResume().
Maybe the number of lines = 0 is the problem. Could be or not? )

+++++++++++++++++++++++++

In Basic.java
```
	public static String getFilePath(String subdir, String subPath) {
		String path = subPath;
		boolean goOn = true;// vv 2017-02-18gt
		if (path != null) {
			if (path.toLowerCase().indexOf("file://") == 0){	
			path = path.substring(7);
			goOn = false;
			}
		}// ^^ 2017-02-18gt
		// if (!subPath.startsWith("/"))	uncomment to enable absolute paths
		if (goOn)// 2017-02-18gt
		{
			StringBuilder bldPath = new StringBuilder(filePath);
			if (subdir != null) { bldPath.append(File.separatorChar).append(subdir); }
			if (subPath != null) { bldPath.append(File.separatorChar).append(subPath); }
			path = bldPath.toString();
		}
		File file = new File(path);
		try                  { path = file.getCanonicalPath(); }
		catch(IOException e) { path = file.getAbsolutePath(); }
		return path;
	}
```

You see "file://" is now supported. So

```
	public static void clearProgram() {
		// Run.running_bas = "";//2017-02-21nm // deleted, because you can use file:// //2017-02-23gt
		lines = new ArrayList<Run.ProgramLine>();			// The lines array list is the program
		lines.add(new Run.ProgramLine(""));					// add an empty string to lines
		Editor.DisplayText="REM Start of BASIC! Program\n";	// Display text is the editors program storage for display
	}

```

```
	public static int loadProgramFileToList(boolean isFullPath, String path, ArrayList<String> list) {

		// MES 2013/01/13 - We got in a jam when moving resources to assets.
		// Files and assets have a different base path. For backward compatibility,
		// we're letting callers use the full path if they only want a file.
		// If a caller wants to use an asset "file", it must set isFullPath false
		// and use a path relative to "source" so getBufferedReader can find the asset.

		int size = 0;
		BufferedReader buf = null;
		String dir = isFullPath ? null : Basic.SOURCE_DIR;
		try { buf = getBufferedReader(dir, path, Basic.Encryption.ENABLE_DECRYPTION); }
		catch (Exception e) { return size; }
		/* deleted, because you can use file:// //2017-02-23gt
		// Save running bas path after making it relative to rfo-basic/source//vv2017-02-21nm
		Run.running_bas = getRelativePath(isFullPath ? path : Basic.getSourcePath(path), Basic.getSourcePath(null));
		//^^2017-02-21nm
		 */
		// Read the file. Insert the the lines into the ArrayList.
		String data = null;

```
/Gregor

**Comment by jMarcS** [comment]
Released in Version 01.91, 2017/03/14
